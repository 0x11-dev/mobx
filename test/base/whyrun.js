"use strict"

const mobx = require("../../src/mobx.ts")
const noop = () => {}

test("whyrun", () => {
    mobx._resetGlobalState()
    const baselog = console.log
    let lastButOneLine = ""
    let lastLine = ""

    const whyRun = function() {
        lastButOneLine = lastLine
        console.log = noop
        lastLine = mobx.whyRun.apply(null, arguments)
        console.log = baselog
        return lastLine
    }

    const x = mobx.observable({
        firstname: "Michel",
        lastname: "Weststrate",
        get fullname() {
            var res = this.firstname + " " + this.lastname
            whyRun()
            return res
        }
    })

    x.fullname
    // TODO: enable this assertion
    // t.ok(lastLine.match(/suspended/), "just accessed fullname"); // no normal report, just a notification that nothing is being derived atm

    expect(whyRun(x, "fullname")).toMatchSnapshot()

    const d = mobx.autorun("loggerzz", () => {
        x.fullname
        whyRun()
    })

    expect(lastButOneLine).toMatchSnapshot()

    expect(lastLine).toMatchSnapshot()

    expect(whyRun(x, "fullname")).toMatchSnapshot()

    expect(whyRun(d)).toMatchSnapshot()

    mobx.transaction(() => {
        x.firstname = "Veria"
        expect(whyRun(x, "fullname").match(/\[idle\]/)).toBeTruthy()

        expect(whyRun(d).match(/\[scheduled\]/)).toBeTruthy()
    })

    expect(lastButOneLine).toMatchSnapshot()

    expect(lastLine).toMatchSnapshot()

    expect(whyRun(x, "fullname")).toMatchSnapshot()

    expect(whyRun(d)).toMatchSnapshot()

    d()

    expect(whyRun(d)).toMatchSnapshot()
    expect(whyRun(x, "fullname")).toMatchSnapshot()
})
