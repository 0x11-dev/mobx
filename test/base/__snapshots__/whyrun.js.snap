// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`trace 1`] = `
Array [
  "- INIT -",
  Arguments [
    "[mobx.trace] 'ObservableObject@4.fullname' tracing enabled",
  ],
  "- SECOND READ -",
  Arguments [
    "[mobx.trace] 'ObservableObject@4.fullname' is being read outside a reactive context and doing a full recompute",
  ],
  "- REACTION -",
  Arguments [
    "[mobx.trace] 'loggerzz' tracing enabled",
  ],
  "- CHANGE -",
  Arguments [
    "[mobx.trace] 'ObservableObject@4.fullname' is invalidated due to a change in: 'ObservableObject@4.firstname'",
  ],
  Arguments [
    "[mobx.trace] 'loggerzz' is invalidated due to a change in: 'ObservableObject@4.fullname'",
  ],
  "- DISPOSE -",
]
`;
